File: plugins/claude-swarm/python/orchestrator.py
"""
Claude Swarm - Orchestrator
Agente orquestrador que coordena todos os workers.
"""

import asyncio
import json
from typing import Dict, Any, List, Optional
from dataclasses import dataclass
from enum import Enum

from anthropic import Anthropic

from config import config, agent_config
from message_broker import broker, Message, MessageType, Priority


class ExecutionStrategy(str, Enum):
    """Estrategias de execucao."""
    FAN_OUT = "fan-out"
    PIPELINE = "pipeline"
    MAP_REDUCE = "map-reduce"
    AUTO = "auto"


@dataclass
class TaskResult:
    """Resultado de uma tarefa."""
    task_id: str
    worker: str
    success: bool
    result: Dict[str, Any]
    duration_ms: int


class Orchestrator:
    """
    Agente Orchestrator do Swarm.
    Coordena workers e executa estrategias de orquestracao.
    """

    def __init__(self):
        self.agent_id = "orchestrator"
        self.config = agent_config.get_config("orchestrator")
        self.client = Anthropic(api_key=config.anthropic_api_key)
        self._running = False
        self._pending_tasks: Dict[str, dict] = {}

    @property
    def system_prompt(self) -> str:
        return """Voce e o Orchestrator do Claude Swarm - o lider que coordena uma equipe de agentes especializados.

Sua responsabilidade:
1. Analisar tarefas complexas e decompor em subtarefas
2. Delegar trabalho para workers apropriados
3. Coordenar execucao paralela ou sequencial
4. Sintetizar resultados de multiplos workers
5. Garantir qualidade e completude das entregas

Workers disponiveis:
- analyst: Analise de codigo, arquitetura e requisitos
- coder: Implementacao e codificacao
- reviewer: Revisao de codigo e qualidade
- tester: Criacao e execucao de testes
- researcher: Pesquisa e coleta de informacoes

Estrategias de execucao:
- fan-out: Todos workers em paralelo (para tarefas independentes)
- pipeline: Workers em sequencia (cada um recebe output do anterior)
- map-reduce: Dividir dados, processar paralelo, agregar

Diretrizes:
- Escolha a estrategia mais eficiente para cada tarefa
- Distribua trabalho de forma balanceada
- Sintetize resultados de forma clara e acionavel
- Identifique e resolva conflitos entre workers"""

    async def _analyze_task(self, description: str) -> Dict[str, Any]:
        """
        Analisa tarefa e determina estrategia e workers.

        Args:
            description: Descricao da tarefa

        Returns:
            Plano de execucao
        """
        response = self.client.messages.create(
            model=self.config["model"],
            max_tokens=2048,
            system="""Voce e um planejador de tarefas. Analise a tarefa e retorne um JSON com:
{
    "strategy": "fan-out" | "pipeline" | "map-reduce",
    "workers": ["lista", "de", "workers", "necessarios"],
    "subtasks": [{"worker": "tipo", "instruction": "instrucao especifica"}],
    "rationale": "explicacao da escolha"
}

Workers disponiveis: analyst, coder, reviewer, tester, researcher

Retorne APENAS o JSON, sem markdown ou explicacoes adicionais.""",
            messages=[{"role": "user", "content": f"Tarefa: {description}"}]
        )

        # Extrair JSON da resposta
        text = response.content[0].text
        try:
            # Tentar parse direto
            return json.loads(text)
        except json.JSONDecodeError:
            # Tentar extrair JSON do texto
            import re
            match = re.search(r'\{.*\}', text, re.DOTALL)
            if match:
                return json.loads(match.group())
            # Fallback
            return {
                "strategy": "fan-out",
                "workers": ["analyst"],
                "subtasks": [{"worker": "analyst", "instruction": description}],
                "rationale": "Fallback para analise simples"
            }

    async def execute_fan_out(
        self,
        description: str,
        workers: List[str],
        subtasks: List[dict]
    ) -> Dict[str, Any]:
        """
        Executa estrategia fan-out (paralelo).

        Args:
            description: Descricao original
            workers: Lista de workers
            subtasks: Subtarefas

        Returns:
            Resultado agregado
        """
        # Verificar workers disponiveis
        health = await broker.health_check()
        alive_workers = [w["name"].replace("worker-", "") for w in health["workers"] if w["status"] == "alive"]

        # Filtrar subtasks para workers disponiveis
        valid_subtasks = [st for st in subtasks if st["worker"] in alive_workers]

        if not valid_subtasks:
            return {"error": "Nenhum worker disponivel", "available": alive_workers}

        # Enviar todas as tarefas em paralelo
        task_ids = []
        for subtask in valid_subtasks:
            message = Message(
                type=MessageType.TASK,
                to_agent=f"worker-{subtask['worker']}",
                payload={
                    "instruction": f"[FAN-OUT] {subtask['instruction']}\n\nTarefa original: {description}",
                    "strategy": "fan-out"
                },
                metadata={"priority": Priority.MEDIUM.value, "ttl": 300000}
            )

            await broker.publish(f"worker-{subtask['worker']}", message)
            task_ids.append({"id": message.id, "worker": subtask["worker"]})
            self._pending_tasks[message.id] = {"worker": subtask["worker"], "status": "pending"}

        # Coletar resultados
        results = []
        for task in task_ids:
            result = await broker.collect(task["id"], timeout_seconds=config.collect_timeout)
            if result:
                results.append(TaskResult(
                    task_id=task["id"],
                    worker=task["worker"],
                    success=True,
                    result=result.payload.get("result", {}),
                    duration_ms=0
                ))
            else:
                results.append(TaskResult(
                    task_id=task["id"],
                    worker=task["worker"],
                    success=False,
                    result={"error": "timeout"},
                    duration_ms=0
                ))

        # Sintetizar
        return await self._synthesize_results(description, results, "fan-out")

    async def execute_pipeline(
        self,
        description: str,
        workers: List[str],
        subtasks: List[dict]
    ) -> Dict[str, Any]:
        """
        Executa estrategia pipeline (sequencial).

        Args:
            description: Descricao original
            workers: Lista de workers em ordem
            subtasks: Subtarefas

        Returns:
            Resultado final
        """
        accumulated_context = description
        results = []

        for i, subtask in enumerate(subtasks):
            instruction = f"""[PIPELINE STAGE {i + 1}/{len(subtasks)}]

{subtask['instruction']}

Contexto acumulado dos estagios anteriores:
{accumulated_context}

Execute sua parte e retorne o resultado para o proximo estagio."""

            message = Message(
                type=MessageType.TASK,
                to_agent=f"worker-{subtask['worker']}",
                payload={"instruction": instruction, "strategy": "pipeline", "stage": i},
                metadata={"priority": Priority.HIGH.value, "ttl": 300000}
            )

            await broker.publish(f"worker-{subtask['worker']}", message)

            result = await broker.collect(message.id, timeout_seconds=config.collect_timeout * 2)

            if result:
                result_data = result.payload.get("result", {})
                results.append(TaskResult(
                    task_id=message.id,
                    worker=subtask["worker"],
                    success=True,
                    result=result_data,
                    duration_ms=0
                ))
                # Acumular contexto
                accumulated_context += f"\n\n[OUTPUT {subtask['worker'].upper()}]:\n{json.dumps(result_data, indent=2)}"
            else:
                results.append(TaskResult(
                    task_id=message.id,
                    worker=subtask["worker"],
                    success=False,
                    result={"error": "timeout"},
                    duration_ms=0
                ))
                break  # Pipeline para no primeiro erro

        return await self._synthesize_results(description, results, "pipeline")

    async def execute_map_reduce(
        self,
        description: str,
        data: Any,
        worker_type: str = "analyst"
    ) -> Dict[str, Any]:
        """
        Executa estrategia map-reduce.

        Args:
            description: Descricao da tarefa
            data: Dados a processar
            worker_type: Tipo de worker para processamento

        Returns:
            Resultado agregado
        """
        # Dividir dados em chunks
        if isinstance(data, list):
            chunk_size = max(1, len(data) // 5)  # 5 chunks
            chunks = [data[i:i+chunk_size] for i in range(0, len(data), chunk_size)]
        elif isinstance(data, str):
            lines = data.split('\n')
            chunk_size = max(1, len(lines) // 5)
            chunks = ['\n'.join(lines[i:i+chunk_size]) for i in range(0, len(lines), chunk_size)]
        else:
            chunks = [data]

        # Map: processar chunks em paralelo
        task_ids = []
        for i, chunk in enumerate(chunks):
            instruction = f"""[MAP-REDUCE CHUNK {i + 1}/{len(chunks)}]

Tarefa: {description}

Processe este chunk de dados:
{json.dumps(chunk) if not isinstance(chunk, str) else chunk}

Retorne resultado estruturado em JSON."""

            message = Message(
                type=MessageType.TASK,
                to_agent=f"worker-{worker_type}",
                payload={"instruction": instruction, "strategy": "map-reduce", "chunk": i},
                metadata={"priority": Priority.MEDIUM.value, "ttl": 300000}
            )

            await broker.publish(f"worker-{worker_type}", message)
            task_ids.append({"id": message.id, "chunk": i})

        # Coletar resultados
        chunk_results = []
        for task in task_ids:
            result = await broker.collect(task["id"], timeout_seconds=config.collect_timeout)
            if result:
                chunk_results.append(result.payload.get("result", {}))

        # Reduce: agregar resultados
        return await self._reduce_results(description, chunk_results)

    async def _synthesize_results(
        self,
        description: str,
        results: List[TaskResult],
        strategy: str
    ) -> Dict[str, Any]:
        """
        Sintetiza resultados de multiplos workers.

        Args:
            description: Tarefa original
            results: Resultados dos workers
            strategy: Estrategia usada

        Returns:
            Sintese final
        """
        successful = [r for r in results if r.success]
        failed = [r for r in results if not r.success]

        results_text = "\n\n".join([
            f"[{r.worker.upper()}]:\n{json.dumps(r.result, indent=2)}"
            for r in successful
        ])

        response = self.client.messages.create(
            model=self.config["model"],
            max_tokens=4096,
            system=self.system_prompt,
            messages=[{
                "role": "user",
                "content": f"""Sintetize os resultados dos workers.

TAREFA ORIGINAL:
{description}

ESTRATEGIA: {strategy}

RESULTADOS DOS WORKERS:
{results_text}

WORKERS QUE FALHARAM: {[r.worker for r in failed] if failed else 'Nenhum'}

Crie uma sintese consolidada com:
1. Resumo executivo
2. Principais descobertas/entregas
3. Conflitos ou inconsistencias (se houver)
4. Recomendacoes de proximos passos"""
            }]
        )

        return {
            "success": True,
            "strategy": strategy,
            "workers_consulted": len(results),
            "workers_successful": len(successful),
            "workers_failed": len(failed),
            "synthesis": response.content[0].text,
            "raw_results": [{"worker": r.worker, "result": r.result} for r in successful]
        }

    async def _reduce_results(
        self,
        description: str,
        chunk_results: List[Dict]
    ) -> Dict[str, Any]:
        """
        Reduz/agrega resultados de map-reduce.

        Args:
            description: Tarefa original
            chunk_results: Resultados dos chunks

        Returns:
            Resultado agregado
        """
        response = self.client.messages.create(
            model=self.config["model"],
            max_tokens=4096,
            system="""Voce e o Reducer do map-reduce. Agregue resultados de multiplos chunks em uma resposta unificada.
Identifique padroes, agregue metricas, e sintetize insights.""",
            messages=[{
                "role": "user",
                "content": f"""TAREFA: {description}

RESULTADOS DOS CHUNKS ({len(chunk_results)}):
{json.dumps(chunk_results, indent=2)}

Agregue todos os resultados em uma resposta unificada."""
            }]
        )

        return {
            "success": True,
            "strategy": "map-reduce",
            "chunks_processed": len(chunk_results),
            "aggregated_result": response.content[0].text
        }

    async def execute(
        self,
        description: str,
        strategy: ExecutionStrategy = ExecutionStrategy.AUTO,
        data: Optional[Any] = None
    ) -> Dict[str, Any]:
        """
        Executa uma tarefa com a estrategia especificada.

        Args:
            description: Descricao da tarefa
            strategy: Estrategia de execucao
            data: Dados opcionais para map-reduce

        Returns:
            Resultado da execucao
        """
        await broker.connect()

        # Analisar tarefa se estrategia for AUTO
        if strategy == ExecutionStrategy.AUTO:
            plan = await self._analyze_task(description)
            strategy = ExecutionStrategy(plan.get("strategy", "fan-out"))
            workers = plan.get("workers", ["analyst"])
            subtasks = plan.get("subtasks", [{"worker": "analyst", "instruction": description}])
        else:
            # Usar analise simples
            plan = await self._analyze_task(description)
            workers = plan.get("workers", ["analyst"])
            subtasks = plan.get("subtasks", [{"worker": "analyst", "instruction": description}])

        print(f"[ORCHESTRATOR] Executando com estrategia: {strategy.value}")
        print(f"[ORCHESTRATOR] Workers: {workers}")

        # Executar estrategia
        if strategy == ExecutionStrategy.FAN_OUT:
            return await self.execute_fan_out(description, workers, subtasks)
        elif strategy == ExecutionStrategy.PIPELINE:
            return await self.execute_pipeline(description, workers, subtasks)
        elif strategy == ExecutionStrategy.MAP_REDUCE:
            return await self.execute_map_reduce(description, data or description)
        else:
            return {"error": f"Estrategia desconhecida: {strategy}"}

    async def broadcast(self, action: str, message: str = "") -> str:
        """
        Envia broadcast para todos workers.

        Args:
            action: Acao
            message: Mensagem

        Returns:
            ID do broadcast
        """
        return await broker.broadcast(action, message)

    async def health_check(self) -> Dict[str, Any]:
        """
        Verifica saude do swarm.

        Returns:
            Status dos workers
        """
        await broker.connect()
        return await broker.health_check()

    async def shutdown(self) -> None:
        """Encerra o swarm graciosamente."""
        print("[ORCHESTRATOR] Iniciando shutdown do swarm...")
        await self.broadcast("shutdown", "Swarm encerrando")
        await asyncio.sleep(2)  # Dar tempo para workers processarem
        await broker.disconnect()
        print("[ORCHESTRATOR] Swarm encerrado")

File: plugins/claude-swarm/typescript/src/orchestrator.ts
/**
 * Claude Swarm - Orchestrator
 * Agente orquestrador que coordena todos os workers.
 */

import Anthropic from '@anthropic-ai/sdk';
import { config, getAgentConfig } from './config';
import { broker } from './message-broker';
import {
  Message,
  MessageType,
  Priority,
  ExecutionStrategy,
  ExecutionPlan,
  ExecutionResult,
  TaskResult,
  Subtask,
} from './types';

export class Orchestrator {
  private agentId = 'orchestrator';
  private config = getAgentConfig('orchestrator');
  private client: Anthropic;
  private pendingTasks: Map<string, { worker: string; status: string }> = new Map();

  constructor() {
    this.client = new Anthropic({ apiKey: config.anthropicApiKey });
  }

  get systemPrompt(): string {
    return `Voce e o Orchestrator do Claude Swarm - o lider que coordena uma equipe de agentes especializados.

Sua responsabilidade:
1. Analisar tarefas complexas e decompor em subtarefas
2. Delegar trabalho para workers apropriados
3. Coordenar execucao paralela ou sequencial
4. Sintetizar resultados de multiplos workers
5. Garantir qualidade e completude das entregas

Workers disponiveis:
- analyst: Analise de codigo, arquitetura e requisitos
- coder: Implementacao e codificacao
- reviewer: Revisao de codigo e qualidade
- tester: Criacao e execucao de testes
- researcher: Pesquisa e coleta de informacoes

Estrategias de execucao:
- fan-out: Todos workers em paralelo (para tarefas independentes)
- pipeline: Workers em sequencia (cada um recebe output do anterior)
- map-reduce: Dividir dados, processar paralelo, agregar

Diretrizes:
- Escolha a estrategia mais eficiente para cada tarefa
- Distribua trabalho de forma balanceada
- Sintetize resultados de forma clara e acionavel
- Identifique e resolva conflitos entre workers`;
  }

  private async analyzeTask(description: string): Promise<ExecutionPlan> {
    const response = await this.client.messages.create({
      model: this.config.model,
      max_tokens: 2048,
      system: `Voce e um planejador de tarefas. Analise a tarefa e retorne um JSON com:
{
    "strategy": "fan-out" | "pipeline" | "map-reduce",
    "workers": ["lista", "de", "workers", "necessarios"],
    "subtasks": [{"worker": "tipo", "instruction": "instrucao especifica"}],
    "rationale": "explicacao da escolha"
}

Workers disponiveis: analyst, coder, reviewer, tester, researcher

Retorne APENAS o JSON, sem markdown ou explicacoes adicionais.`,
      messages: [{ role: 'user', content: `Tarefa: ${description}` }],
    });

    const text = (response.content[0] as Anthropic.TextBlock).text;
    try {
      return JSON.parse(text);
    } catch {
      const match = text.match(/\{.*\}/s);
      if (match) {
        return JSON.parse(match[0]);
      }
      return {
        strategy: ExecutionStrategy.FAN_OUT,
        workers: ['analyst'],
        subtasks: [{ worker: 'analyst', instruction: description }],
        rationale: 'Fallback para analise simples',
      };
    }
  }

  async executeFanOut(
    description: string,
    workers: string[],
    subtasks: Subtask[]
  ): Promise<ExecutionResult> {
    const health = await broker.healthCheck();
    const aliveWorkers = health.workers
      .filter((w) => w.status === 'alive')
      .map((w) => w.name.replace('worker-', ''));

    const validSubtasks = subtasks.filter((st) => aliveWorkers.includes(st.worker));

    if (validSubtasks.length === 0) {
      return { success: false, strategy: 'fan-out', error: 'Nenhum worker disponivel' };
    }

    const taskIds: Array<{ id: string; worker: string }> = [];

    for (const subtask of validSubtasks) {
      const message = broker.createMessage(
        MessageType.TASK,
        `worker-${subtask.worker}`,
        {
          instruction: `[FAN-OUT] ${subtask.instruction}\n\nTarefa original: ${description}`,
          strategy: 'fan-out',
        },
        Priority.MEDIUM
      );

      await broker.publish(`worker-${subtask.worker}`, message);
      taskIds.push({ id: message.id, worker: subtask.worker });
      this.pendingTasks.set(message.id, { worker: subtask.worker, status: 'pending' });
    }

    const results: TaskResult[] = [];
    for (const task of taskIds) {
      const result = await broker.collect(task.id, config.collectTimeout);
      if (result) {
        results.push({
          taskId: task.id,
          worker: task.worker,
          success: true,
          result: (result.payload.result as Record<string, unknown>) || {},
          durationMs: 0,
        });
      } else {
        results.push({
          taskId: task.id,
          worker: task.worker,
          success: false,
          result: { error: 'timeout' },
          durationMs: 0,
        });
      }
    }

    return this.synthesizeResults(description, results, 'fan-out');
  }

  async executePipeline(
    description: string,
    workers: string[],
    subtasks: Subtask[]
  ): Promise<ExecutionResult> {
    let accumulatedContext = description;
    const results: TaskResult[] = [];

    for (let i = 0; i < subtasks.length; i++) {
      const subtask = subtasks[i];
      const instruction = `[PIPELINE STAGE ${i + 1}/${subtasks.length}]

${subtask.instruction}

Contexto acumulado dos estagios anteriores:
${accumulatedContext}

Execute sua parte e retorne o resultado para o proximo estagio.`;

      const message = broker.createMessage(
        MessageType.TASK,
        `worker-${subtask.worker}`,
        { instruction, strategy: 'pipeline', stage: i },
        Priority.HIGH
      );

      await broker.publish(`worker-${subtask.worker}`, message);

      const result = await broker.collect(message.id, config.collectTimeout * 2);

      if (result) {
        const resultData = (result.payload.result as Record<string, unknown>) || {};
        results.push({
          taskId: message.id,
          worker: subtask.worker,
          success: true,
          result: resultData,
          durationMs: 0,
        });
        accumulatedContext += `\n\n[OUTPUT ${subtask.worker.toUpperCase()}]:\n${JSON.stringify(resultData, null, 2)}`;
      } else {
        results.push({
          taskId: message.id,
          worker: subtask.worker,
          success: false,
          result: { error: 'timeout' },
          durationMs: 0,
        });
        break;
      }
    }

    return this.synthesizeResults(description, results, 'pipeline');
  }

  async executeMapReduce(
    description: string,
    data: unknown,
    workerType: string = 'analyst'
  ): Promise<ExecutionResult> {
    let chunks: unknown[];

    if (Array.isArray(data)) {
      const chunkSize = Math.max(1, Math.ceil(data.length / 5));
      chunks = [];
      for (let i = 0; i < data.length; i += chunkSize) {
        chunks.push(data.slice(i, i + chunkSize));
      }
    } else if (typeof data === 'string') {
      const lines = data.split('\n');
      const chunkSize = Math.max(1, Math.ceil(lines.length / 5));
      chunks = [];
      for (let i = 0; i < lines.length; i += chunkSize) {
        chunks.push(lines.slice(i, i + chunkSize).join('\n'));
      }
    } else {
      chunks = [data];
    }

    const taskIds: Array<{ id: string; chunk: number }> = [];

    for (let i = 0; i < chunks.length; i++) {
      const instruction = `[MAP-REDUCE CHUNK ${i + 1}/${chunks.length}]

Tarefa: ${description}

Processe este chunk de dados:
${typeof chunks[i] === 'string' ? chunks[i] : JSON.stringify(chunks[i])}

Retorne resultado estruturado em JSON.`;

      const message = broker.createMessage(
        MessageType.TASK,
        `worker-${workerType}`,
        { instruction, strategy: 'map-reduce', chunk: i },
        Priority.MEDIUM
      );

      await broker.publish(`worker-${workerType}`, message);
      taskIds.push({ id: message.id, chunk: i });
    }

    const chunkResults: Record<string, unknown>[] = [];
    for (const task of taskIds) {
      const result = await broker.collect(task.id, config.collectTimeout);
      if (result) {
        chunkResults.push((result.payload.result as Record<string, unknown>) || {});
      }
    }

    return this.reduceResults(description, chunkResults);
  }

  private async synthesizeResults(
    description: string,
    results: TaskResult[],
    strategy: string
  ): Promise<ExecutionResult> {
    const successful = results.filter((r) => r.success);
    const failed = results.filter((r) => !r.success);

    const resultsText = successful
      .map((r) => `[${r.worker.toUpperCase()}]:\n${JSON.stringify(r.result, null, 2)}`)
      .join('\n\n');

    const response = await this.client.messages.create({
      model: this.config.model,
      max_tokens: 4096,
      system: this.systemPrompt,
      messages: [
        {
          role: 'user',
          content: `Sintetize os resultados dos workers.

TAREFA ORIGINAL:
${description}

ESTRATEGIA: ${strategy}

RESULTADOS DOS WORKERS:
${resultsText}

WORKERS QUE FALHARAM: ${failed.length > 0 ? failed.map((r) => r.worker).join(', ') : 'Nenhum'}

Crie uma sintese consolidada com:
1. Resumo executivo
2. Principais descobertas/entregas
3. Conflitos ou inconsistencias (se houver)
4. Recomendacoes de proximos passos`,
        },
      ],
    });

    return {
      success: true,
      strategy,
      workersConsulted: results.length,
      workersSuccessful: successful.length,
      workersFailed: failed.length,
      synthesis: (response.content[0] as Anthropic.TextBlock).text,
      rawResults: successful.map((r) => ({ worker: r.worker, result: r.result })),
    };
  }

  private async reduceResults(
    description: string,
    chunkResults: Record<string, unknown>[]
  ): Promise<ExecutionResult> {
    const response = await this.client.messages.create({
      model: this.config.model,
      max_tokens: 4096,
      system: `Voce e o Reducer do map-reduce. Agregue resultados de multiplos chunks em uma resposta unificada.
Identifique padroes, agregue metricas, e sintetize insights.`,
      messages: [
        {
          role: 'user',
          content: `TAREFA: ${description}

RESULTADOS DOS CHUNKS (${chunkResults.length}):
${JSON.stringify(chunkResults, null, 2)}

Agregue todos os resultados em uma resposta unificada.`,
        },
      ],
    });

    return {
      success: true,
      strategy: 'map-reduce',
      synthesis: (response.content[0] as Anthropic.TextBlock).text,
    };
  }

  async execute(
    description: string,
    strategy: ExecutionStrategy = ExecutionStrategy.AUTO,
    data?: unknown
  ): Promise<ExecutionResult> {
    await broker.connect();

    let plan: ExecutionPlan;
    let effectiveStrategy = strategy;

    if (strategy === ExecutionStrategy.AUTO) {
      plan = await this.analyzeTask(description);
      effectiveStrategy = plan.strategy as ExecutionStrategy;
    } else {
      plan = await this.analyzeTask(description);
    }

    console.log(`[ORCHESTRATOR] Executando com estrategia: ${effectiveStrategy}`);
    console.log(`[ORCHESTRATOR] Workers: ${plan.workers.join(', ')}`);

    switch (effectiveStrategy) {
      case ExecutionStrategy.FAN_OUT:
        return this.executeFanOut(description, plan.workers, plan.subtasks);
      case ExecutionStrategy.PIPELINE:
        return this.executePipeline(description, plan.workers, plan.subtasks);
      case ExecutionStrategy.MAP_REDUCE:
        return this.executeMapReduce(description, data || description);
      default:
        return { success: false, strategy: String(effectiveStrategy), error: 'Estrategia desconhecida' };
    }
  }

  async broadcast(action: string, message: string = ''): Promise<string> {
    return broker.broadcast(action, message);
  }

  async healthCheck(): Promise<ReturnType<typeof broker.healthCheck>> {
    await broker.connect();
    return broker.healthCheck();
  }

  async shutdown(): Promise<void> {
    console.log('[ORCHESTRATOR] Iniciando shutdown do swarm...');
    await this.broadcast('shutdown', 'Swarm encerrando');
    await new Promise((resolve) => setTimeout(resolve, 2000));
    await broker.disconnect();
    console.log('[ORCHESTRATOR] Swarm encerrado');
  }
}
